### PR TITLE
Make BuildPackages.sh to accept packages as arguments, log into file and color the output

### PR DESCRIPTION
**Pull request by hungaborhorvath**
Log every package to different file with Buildpackages.sh fixing #1014, and handle packages given as arguments

The first argument should give the gaproot directory with ````--with-gaproot=````. All further arguments are considered package directories (their names have to be complete!). If first argument does not start with ````--with-gaproot=````, then it is also considered a package to be built. If no packages are given as arguments, then everything is built.

For better readability, added an extra empty line before checking a new package, 
and also before reporting that a package is failed to compile. This latter line is colored by yellow and preceded by ````WARNING````. Errors are red, text written by the script is green. 

Packages failed to build are logged to ````pkg/log/fail.log````, together with a
date of when they failed.

Every package ````PKG```` build output is logged to a file in ````pkg/log/$PKG.log````.
The whole output is written into ````pkg/log/buildpackages.log````.

If there is ````stderr```` output, it is written to ````.err```` files, then the ````stdout````
output is written to ````.out```` files. The ````.log```` files contain the whole output
in sequence. Every file contains the _colored_ text, ie. ````cat```` understands it and will print to the screen in a colored way. 

Every ````.log````, ````.out````, ````.err```` file is overwritten at every rerun of
````buildpackages.sh````, ***except*** for ````fail.log````, which is always appended
together with the date of the run.

Also, ````date```` is run and logged into every package log file to keep some track of when happened whatever happened.

Relative paths are replaced by absolute paths. 

Output of ````ls```` is dangerous to parse, replaced by ````shopt -s nullglob```` where appropriate.

Building carat is fixed.

### CODE REVIEW

**Comment by ChrisJefferson** [comment]
These changes seem simple and uncontroversial to me. I will just see if anyone else wants to comment.

**Comment by hungaborhorvath** [comment]
@ChrisJefferson I have updated the script, now it logs to different files in a more or less reasonable manner. Also, names of failed packages are logged into ````pkg/log/fail.log````

**Comment by ChrisJefferson** [comment]
I am now happy with this, I'll just see if anyone else wants to comment.

**Comment by fingolfin** [comment]
I'll try it on my OS X 10.11 system. @hungaborhorvath on what kind of OS / machine did you test it?

File: bin/BuildPackages.sh [hunk 1]
```diff
@@ -1,5 +1,15 @@
 #!/usr/bin/env bash
 
+LOGDIR=log
+if [ ! -d "$LOGDIR" ]
+  then
+  mkdir -p $LOGDIR
+fi
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: The check is redundant: just replace this by `mkdir -p`

File: bin/BuildPackages.sh [hunk 2]
```diff
@@ -1,5 +1,15 @@
...
+LOGFILE=buildpackages
+FAILPKGFILE=fail
+
+build_all_packages() {
+
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: This change breaks handling of the argument. I have my `pkg` outside the repository, so that I can share it with multiple conccurrent builds. I invoke the script like this:
```
$GAPROOT/bin/BuildPackages.sh $GAPROOT
```
This is broken with this PR. You need to parse arguments to the script first. Then, pass GAPDIR as a mandatory argument to `build_all_packages`.

File: bin/BuildPackages.sh [hunk 3]
```diff
@@ -1,5 +1,15 @@
...
 set -e
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: Hmm, why is the `set -e` here -- why not have it active for the whole script? I.e. change the shebang to
```
#!/usr/bin/env bash -e
 `

File: bin/BuildPackages.sh [hunk 4]
```diff
@@ -39,7 +49,9 @@ if [ -f gapicon.bmp ]
 fi
 
 # We need any subdirectory, to test if $GAPDIR is right
-SUBDIR=`ls -d */ | head -n 1`
+# dangerous to parse output of ls
+#SUBDIR=`ls -d */ | head -n 1`
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: It's pointless to keep the old call to `ls` and the warning here -- remove them.

File: bin/BuildPackages.sh [hunk 5]
```diff
@@ -39,7 +49,9 @@ if [ -f gapicon.bmp ]
...
+SUBDIR=`find -L ./ -maxdepth 1 -mindepth 1 -type d | sed 's|./||' | head -n 1`
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: `find` may not be available on all systems, but OK, we can risk that. But in a regex, `.` has a special meaning, so strictly speaking you should escape it in the `sed` invocation.

But in fact, if you replace `./` by `.` in the `find` invocation, you can just drop the call to `sed`...

Comment (fingolfin) [inline bin/BuildPackages.sh]: I realize that you are just adapting existing code here, so I'll accept this. But I actually think it's not a great solution. Better would be to convert any relative paths into absolute ones ... in particular, don't require the user to specify a path "relative from a package directory. Instead, require the user to specify a path which is valid from the place `BuildPackages.sh` is called (i.e. inside some `pkg` directory). Then, if that path is relative, convert it to an absolute path. And then use that. So in particular, we'd not default to the GAPDIR `../..` but rather `..`. 

But again, I am not asking you to make that change now, just pointint this out.

File: bin/BuildPackages.sh [hunk 6]
```diff
@@ -85,6 +98,7 @@ make TOPDIR=`pwd`
 chmod -R a+rX .
 cd bin
 aa=`./config.guess`
+# TODO dangerous to parse output of ls, use find instead
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: So let's resolve this TODO... First off, `aa` is a horrible name. Better would be something like `hostarch`... But even that is problematic, because the `config.guess` we are invoking here might differ from the one GAP uses. The correct solution would be 
```sh
source "$GAPDIR/sysinfo.gap
```
and then use the variable `GAParch` set by it. Then we can simply refer to `$GAPDIR/bin/$GAParch` and need neither `ls` nor `find`.

Comment (fingolfin) [inline bin/BuildPackages.sh]: But I still don't quite understand what the `ln -s` below is meant to do... What is it needed for? Perhaps there is a better solution...

File: bin/BuildPackages.sh [hunk 7]
```diff
@@ -186,8 +208,33 @@ do
...
+        then
+        rm -f "$LOGDIR/$dir.err"
+        rm -f "$LOGDIR/$dir.out"
+      fi
     else
       echo "$dir is not a GAP package -- no PackageInfo.g"
-    fi;
-done;
+    fi
+done < <(find -L ./ -maxdepth 1 -mindepth 1 -type d | sed 's|./||')
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: Again, just change `./` to `.` and get rid of the sed ?

Comment (fingolfin) [inline bin/BuildPackages.sh]: Actually, why do it like this, instead of the more common
```shell
for dir in `find -L . -maxdepth 1 -mindepth 1 -type d` ;do
...
```

File: bin/BuildPackages.sh [hunk 8]
```diff
@@ -186,8 +208,33 @@ do
...
+}
+
+# Log error to .err, output to .out, everything to .log
+( build_all_packages > >(tee $LOGDIR/$LOGFILE.out) 2> >(tee $LOGDIR/$LOGFILE.err >&2) )> >( tee $LOGDIR/$LOGFILE.log ) 2>&1
+
+# remove superfluous buildpackages log files if there was no error message
+if [ ! -s "$LOGDIR/$LOGFILE.err" ]
+  then
+  rm -f "$LOGDIR/$LOGFILE.err"
+  rm -f "$LOGDIR/$LOGFILE.out"
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: Why delete the .out here? Even if there was no error *detected*, I still might want to use the log (e.g. to debug why a failure was not detected as such ;-)

File: bin/BuildPackages.sh (hunk 5 above)
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: I use debian stable, that comes with bash 4.3 (is ````find```` not a bash command?). For me ````find -L .```` gives the directories, but all of them are preceded by ````./````, which does not matter here, but does matter later. So I do not see how ````sed```` can be dropped. Maybe I can find some (other) bash trick to do that. 

Further, I do not understand why I should escape ````.```` in the ````sed```` invocation, where that is already in single quotes, so escaped, right? I must be missing something elementary here.

File: bin/BuildPackages.sh (hunk 6 above)
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: The reason I did not resolve this is because I did not check properly how to build carat. It failed on my system, and did not investigate further, yet.

**Comment by hungaborhorvath** [comment]
@fingolfin Thank you for your suggestions, I will change the script. I replied to some of your comments. BTW, I use debian stable (currently jessie).

File: bin/BuildPackages.sh (hunk 5 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: No, `find` is not a bash built-in (just type `help` or `help find` to get a list of built-in commands)

And `find -L ./` actually gives me dir names prefixed with `.//`, so with the grep, they still have a `/` prefix. 

But regardless of whether the prefix is `./` or `.//`, I don't see why it matters? There is no difference between `cd DIR` and `cd ./DIR` and `cd .//DIR` that I could tell. (Of course I still think that using `SUBDIR` is a HACK to start with, but we can clean that up later).

Finally, escaping: These are two different kinds of escaping :-). Using single quotes (versus double quotes, versus no quotes at all) affects how bash interprets inputs... But in fact, the string `|./||` contains no special characters and no spaces, hence the following three commands are all equialvalanet in bash:
```
sed 's|./||'
sed "s|./||"
sed "s|./||"
```
In each case, the `sed` executable sees as its first argument the same string.

However, the character `.` has a special meaning to sed respectively in regular expressions: It matches *any* character. So `./` matches not just `./` but also `x/` or `//` or ... luckily, in this case it usually won't matter -- unless perhaps a directory name contained a special character, which is unlikely *in this particular situation*. But I recommend coding defensively, instead of relying on unspoken assumptions, and thus sed should be told to match a period, not "an arbitrary" character. Which requires escaping on the level of sed, not on the level of bash. I.e.:
```
sed '|\./||'
```
But again, I think we can just not call `sed` at all here and it would be fine :)

File: bin/BuildPackages.sh (hunk 6 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: OK, fair enough. I'll look into it

File: bin/BuildPackages.sh (hunk 7 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: But that advice only applies to reading from a file, and you are not reading from a file. You are iterating over a list of paths, and you definitely should use a `for` loop for that. (Feel free to use `$(...)$` instead of backticks, though).

File: bin/BuildPackages.sh (hunk 8 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: Ahh, I missed that there are three files to which the script logs. OK, that makes sense then!

**Comment by hungaborhorvath** [comment]
@fingolfin Thank you. I think I have made all the changes you asked for. I also did some cleanup, but I tried to do that in the last commit, to more or less properly document it. I am happy to clean up the carat code if you help me. Or if you have no time, I can try to look into it myself. 

I plan to do some more cleanup, but I can stop if that would make harder to review this PR.

File: bin/BuildPackages.sh (hunk 2 above)
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: Now, ````"$@"```` is parsed to the call of ````build_all_packages````

**Comment by hungaborhorvath** [comment]
Oh, about the absolute paths: I do not think it is really worth the effort. I will think about it a bit, but I doubt that I will do that change.

**Comment by fingolfin** [comment]
I guess the `&&` are from a time when the script was run without `-e` enabled. With `-e`, this is indeed superfluous.

File: bin/BuildPackages.sh [hunk 9]
```diff
@@ -21,46 +29,48 @@ set -e
...
 # for improved compatibility with *BSD systems where "make"
 # is BSD make, not GNU make.
-if ! [ x`which gmake` = "x" ] ; then
-  MAKE=gmake
+if ! [[ "x$(which gmake)" = "x" ]]
+then
+  MAKE="gmake"
 else
-  MAKE=make
+  MAKE="make"
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: Why did you add all those superfluous quotes? I mean, they make sense around paths (which could contain spaces), so in particular for GAPDIR. But for constants where we know that quoting is no issue, this seems a bit iffy...

File: bin/BuildPackages.sh [hunk 10]
```diff
@@ -81,60 +92,75 @@ build_carat() {
...
+for dir in $(find -L . -maxdepth 1 -mindepth 1 -type d)
 do
-    if [ -e $dir/PackageInfo.g ]; then
-      dir="${dir%/}"
+  if [[ -e "$dir/PackageInfo.g" ]]
+  then
+    dir="${dir%/}"
+    dir="${dir##*/}"
+
+    buildpackage() {
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: Defining a function in the middle of a loop seems weird... Would be nice if it was defined before the loop, and took `dir` as argument.

File: bin/BuildPackages.sh [hunk 11]
```diff
@@ -184,10 +210,36 @@ do
...
+      (buildpackage > >(tee "$LOGDIR/$dir.out") 2> >(tee "$LOGDIR/$dir.err" >&2) )> >(tee "$LOGDIR/$dir.log" ) 2>&1
       # end subshell
+      # remove superfluous log files if there was no error message
+      if [[ ! -s "$LOGDIR/$dir.err" ]]
+      then
+        rm -f "$LOGDIR/$dir.err"
+        rm -f "$LOGDIR/$dir.out"
+      fi
     else
       echo "$dir is not a GAP package -- no PackageInfo.g"
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: That message actually bothers me -- I have a `.git` directory in my `pkg` dir, and it prints this "warning"... And now that there is a `log` dir, it'll also print it for that. What purpose does it server? Can't we just keep silent?

File: bin/BuildPackages.sh (hunk 10 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: another advantage of this is that we are then very close to being able to build just a single package -- say, specified via an argument to the script.

I.e. the script could have some actual argument parsing at the start, accepting an `--with-gaproot=DIR` argument, optionally followed by a list of package names (or names of package dirs) to build...

File: bin/BuildPackages.sh (hunk 11 above)
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: I do not mind removing it, but may have served some purpose originally. @ChrisJefferson could you comment on this?

Comment (ChrisJefferson) [inline bin/BuildPackages.sh]: Well, it seemed like the sensible thing to do at the time, but I'm equally happy to silently skip directories with no `PackageInfo.g`, particularly if there is a good reason (and the existence of `.git` / `.hg` directories seems like a good reason).

File: bin/BuildPackages.sh (hunk 9 above)
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: Constants can change if someone rewrites the script. But I do not mind, I will just remove them.

**Comment by hungaborhorvath** [comment]
@fingolfin I think I applied all your suggestions, thanks. I was thinking about your note on modifying the script to accept packages, as well. I think I can do it. Should I do it in this PR then? If yes, then is it enough to log the particular packages? How should we handle the ````fail.log```` file (where we keep failed packages)? 

Further, I was thinking about whether to overwrite or append the logs. Currently we overwrite every log ***except*** ````fail.log```` (where we keep failed packages). This has the advantage that if there are some errors (i.e. existing ````.err```` file), then one can find the problem more easily than if there is only one file and one needs to comb through that, find the last compile, etc. But with failed packages not being overwritten (but a date written into them) one can keep track of when they built packages and what failed then.

**Comment by fingolfin** [comment]
I think handling single packages, and more elaborate logging schemes, should be reserved for another PR; better to ship something working and incrementally improve it, than forever refinging and refining but never deploying...

(That said, I also noticed that a lot of essentially trivial log files are generated, namely for packages which build nothing, and those are somehow distracting -- I'd kind of prefer if we did not generate those at all, as they contain no useful information. But again, there's no need to address this in this PR.)

**Comment by fingolfin** [comment]
LEst I forgot: This script has another major flaw (which it had before this PR, too): It overwrites CFLAGS, CXXFLAGS and LDFLAGS, meaning that the caller has no way to customize them (which I need to do on my system, to ensure packages can use GMP). It would be better if we appended to those FLAGS, instead of overwriting.

And then it would be easy to color the output a bit (as in the notice/warning/error function here: https://github.com/fingolfin/ReleaseTools/blob/master/release), in particular the `==== Checking PACKAGENAME` bits -- that, too, would probably help navigating the scrollback.

**Comment by fingolfin** [comment]
(But again, neither needs to happen in this PR)

**Comment by hungaborhorvath** [comment]
@fingolfin Sorry, I could not help myself, so I made the modifications to handle packages given as parameters. I will shortly update the first post to reflect the new syntax. 

I would very much appreciate if people tested this. So far for me the ````pargap```` and ````anupq```` packages fail to build, and previously it was logged properly, and now it is not, so I will need to look into that. But other than this, on my system it seems to perform ok.

**Comment by hungaborhorvath** [comment]
@fingolfin I fixed the carat build. At least it now builds properly on my system, and GAP can load it. I used your suggestion about reading ````$GAParch```` from ````$GAPDIR/sysinfo.gap````. Old code is removed. Further, it checks if the symlinks already exist, thus not reporting failure where there is none.

File: bin/BuildPackages.sh [hunk 12]
```diff
@@ -21,55 +50,82 @@ set -e
...
 
-# We need any subdirectory, to test if $GAPDIR is right
-SUBDIR=`ls -d */ | head -n 1`
-if ! (cd $SUBDIR && [ -f $GAPDIR/sysinfo.gap ])
-  then
-    echo "$GAPDIR is not the root of a gap installation (no sysinfo.gap)"
-    echo "Please provide the absolute path of your GAP root directory as"
-    echo "first argument to this script."
-    exit 1
+# pakcages to build
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: typo: pakcages

File: bin/BuildPackages.sh [hunk 13]
```diff
@@ -21,55 +50,82 @@ set -e
...
+    --with-gaproot=*)
+      GAPDIR="$(sed 's|--with-gaproot=||' <<< $1)"
+      shift
+    ;;
+
+    *)
+      cd ..
+      GAPDIR="$(pwd)"
+      cd "$CURDIR"
+    ;;
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: I don't understand: So if I pass in any argument that is *not* of the form `---with-gaproot=*`, then that argument is ignored, and you set GAPDIR to the parent directory of the current dir? So if I enter the (previously valid) command sequence
```
cd /my/custom/pkg
$GAPDIR/bin/BuildPackages.sh $GAPDIR
```
it will now break?!

Comment (fingolfin) [inline bin/BuildPackages.sh]: Perhaps this should just be changed to
```
*)
  GAPDIR="$1"
  shift
;;
```

File: bin/BuildPackages.sh [hunk 14]
```diff
@@ -21,55 +50,82 @@ set -e
...
-    exit 1
+# CURDIR
+CURDIR="$(pwd)"
+
+# GAPDIR
+if [[ "$#" -ge 1 ]]
+then
+  case "$1" in
+    --with-gaproot=*)
+      GAPDIR="$(sed 's|--with-gaproot=||' <<< $1)"
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: This will break if $1 contains spaces (i.e. if the user specifies a path including spaces). The correct way then is to write this:
```
GAPDIR=$(sed 's|--with-gaproot=||' <<< "$1")
```
Or write this (not using bash extensions):
```
GAPDIR=$(echo "$1 | sed 's|--with-gaproot=||')
```

File: bin/BuildPackages.sh [hunk 15]
```diff
@@ -79,115 +135,197 @@ build_carat() {
...
 cd bin
-aa=`./config.guess`
-for x in "`ls -d1 $GAPDIR/bin/${aa}*`"; do
- ln -s "$aa" "`basename $x`"
-done
+# This sets GAParch as it should be.
+# Alternatively, one could check $GAPDIR/bin for all directories instead.
+source "$GAPDIR/sysinfo.gap"
+# If the symbolic link does not exist then create it.
+if [[ ! -L ./"$GAParch" ]]
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: Why is there an `./` there? Perhaps as a safe guard against an empty GAParch? But then we end up doing something wrong, instead of aborting, so not doing this would be preferable, or alternatively an explicit check for  GAParch being non-empty.

In short: Remove the `./`?

File: bin/BuildPackages.sh [hunk 16]
```diff
@@ -79,115 +135,197 @@ build_carat() {
...
+  # We assume that there is only one appropriate directory....
+  for archdir in *
+  do
+    if [[ -d "$archdir" ]] && [[ ! -L "$archdir" ]]
+    then
+      ln -s ./"$archdir" ./"$GAParch"
+    fi
+  done
+  shopt -u nullglob
+fi
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: I don't understand what the code above is meant to achieve at all: So it iterates over all existing directories inside `bin` which are not symlinks, and then creates a symlink from them to `$GAParch`? Doesn't that mean we get an error if there is more than one?

But why does this code exist in the first place? What problems does it solve? I am somewhat sceptical it is the right thing to do... but perhaps I am wrong, but then this definitely deserves a comment explaining the rationale here.

File: bin/BuildPackages.sh [hunk 17]
```diff
@@ -21,55 +50,82 @@ set -e
...
+    *)
+      cd ..
+      GAPDIR="$(pwd)"
+      cd "$CURDIR"
+    ;;
+  esac
+else
+  cd ..
+  GAPDIR="$(pwd)"
+  cd "$CURDIR"
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: You might want to use `pushd` and `popd` here instead of `cd`. I.e.
```
pushd ..
GAPDIR=$(pwd$
popd
```

File: bin/BuildPackages.sh (hunk 13 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: Aha, it seems that (reading on) you are simply changing the semantics of this script. I guess I don't terribly mind, but it's not super nice either.. Hrm.

You could make your intent clearer, though, and at the same time reduce code duplication, by restructung the code to look like this:
```sh
# GAPDIR
GAPDIR=
if [[ "$#" -ge 1 ]]
then
  case "$1" in
    --with-gaproot=*)
      GAPDIR=$(sed 's|--with-gaproot=||' <<< "$1")
      shift
    ;;
fi
if [[ -z "$GAPDIR" ]]
then
  pushd ..
  GAPDIR="$(pwd)"
  popd
fi
notice "Using GAP location: $GAPDIR"
```

File: bin/BuildPackages.sh [hunk 18]
```diff
@@ -21,55 +50,82 @@ set -e
...
+else
+  # if there are no arguments, then build all packages
+  shopt -s nullglob
+  PACKAGES=(*)
+  shopt -u nullglob
 fi
 
-if (cd $SUBDIR && grep 'ABI_CFLAGS=-m32' $GAPDIR/Makefile > /dev/null) ; then
-  echo "Building with 32-bit ABI"
+# We need to test if $GAPDIR is right
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: Wouldn't it make more sense to do this right after setting GAPDIR, and before dealing with PACKAGES?

File: bin/BuildPackages.sh [hunk 19]
```diff
@@ -79,115 +135,197 @@ build_carat() {
...
+  echo ""
+  warning "Failed to build $PKG"
+  echo "$PKG" >> "$LOGDIR/$FAILPKGFILE.log"
 }
 
 run_configure_and_make() {
   # We want to know if this is an autoconf configure script
   # or not, without actually executing it!
-  if [ -f autogen.sh ] && ! [ -f configure ] ; then
+  if [[ -f autogen.sh ]] && ! [[ -f configure ]]
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: If you want to full embrace bash (i.e. moving away from a simple sh script), you could use that [[ support &&, and !, and so you could write
```
if [[ -f autogen.sh && ! -f configure ]]
```

File: bin/BuildPackages.sh [hunk 20]
```diff
@@ -79,115 +135,197 @@ build_carat() {
...
+  echo ""
+  date
+  echo ""
+  notice "==== Checking $PKG"
+  (  # start subshell
+  set -e
+  cd "$CURDIR/$PKG"
+  case "$PKG" in
+    anupq*)
+      ./configure "CFLAGS=-m32 LDFLAGS=-m32 --with-gaproot=$GAPDIR" && \
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: I see you retained the `&&` here after all. So do they serve a purpose? In that case, perhaps add a comment explaining it? Otherwise, perhaps remove them after all?

File: bin/BuildPackages.sh [hunk 21]
```diff
@@ -79,115 +135,197 @@ build_carat() {
...
-        ;;
-      esac;
-      ) || build_fail
-      # end subshell
-    else
-      echo "$dir is not a GAP package -- no PackageInfo.g"
-    fi;
-done;
+  PKG="${PKG%/}"
+  PKG="${PKG##*/}"
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: What do these two lines do? Add a comment (not everybody is a bash expert, and this script needs to b maintainable by as many people as possibl).

File: bin/BuildPackages.sh (hunk 20 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: It seems you actually removed them, but added them back in 54c0df3eb2bd747560c without an explanation...

File: bin/BuildPackages.sh (hunk 13 above)
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: Yes, from now on the code
````
cd /my/custom/pkg
$GAPDIR/bin/BuildPackages.sh $GAPDIR
````
does not work. If you do not start the first argument by ````--with-gaproot````, then it is considered to be a package name. Now I added a warning about this into the error message if ````$GAPDIR```` is not found properly.

(To be honest, I did not want to write a very sophisticated check on whether the argument is a proper gap directory or a package or something else. I felt that this was an easier way.)

In any case, I applied all your suggestions here.

File: bin/BuildPackages.sh (hunk 16 above)
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: I will be honest with you, the README of the package is not very clear. I do not really understand why this is not done by ````carat````'s own ````make````. But anyway, in most cases there will be only one directory here, and then a symlink (named ````$GAParch````) will be created to this one directory. I did not use ````ls```` or ````find````, because their output is not entirely reliable (although probably would suffice here).

If there are more directories here, then yes, there will be an error. However, I can only imagine more directories here if somebody is an expert and compiled ````carat```` using several different compilers. But then they should know what they are doing and should not care about ````BuildPackages.sh```` doing the building for this particular package, but rather do the symlinking themselves.

In any case, it would be nice if somebody from ````carat````'s developers/maintainers could comment on this so we could have a more foolproof solution.

**Comment by hungaborhorvath** [comment]
@fingolfin Thank you very much for the feedback! I have applied all your suggestions, and replied to two of your comments. 

I am not sure what to do about colors going into the logs or not. I personally like them, because ````cat```` understands it, but I agree that it can be annoying (e.g. when I view the logs with Midnight Commander's view). In any case, I do not see a very easy solution to removing them from the logs, so if you do not mind, I would leave the colors there....

File: bin/BuildPackages.sh [hunk 22]
```diff
@@ -79,115 +141,203 @@ build_carat() {
...
 # path to some data files burned in.
 zcat carat-2.1b1.tgz | tar pxf -
-ln -s carat-2.1b1/bin bin
+# If the symbolic link does not exist then create it.
+# Does not check if bin exists already with some content or as a file, etc.
+# Thus this code is unstable, but should be corrected by package author
+# and not by this script.
+if [[ ! -L ./bin ]]
+then
+  ln -s ./carat-2.1b1/bin ./bin
```
Comment (fingolfin) [inline bin/BuildPackages.sh]: Why do you have those redundant `./` in there?

File: bin/BuildPackages.sh (hunk 16 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: Well, your new code does something different than the old code! And that code tries to allow people with multiple GAP build configs (a little known and little used feature, I believe) on a single computer to build carat once, and use it with all build configs.

On any given GAP installation, the default build config is named "default", and comes in 32 and 64 bit flavors. If `config.guess` returns `FOOBAR`, then GAP puts the compilation results into `bin/FOOBAR-default64` for 64-bit builds, and `bin/FOOBAR-default32` for 32-bit builds. You might in addition have a "debug" build config, then its binaries would end up in `bin/FOOBAR-debug64` resp. `bin/FOOBAR-debug32`. You can have all these binaries simultaneously, which can be handy if you want to switch quickly between different build settings. Also, if your GAP root is on a network share, it might be used by another machine with a different architecture, say `QUUX`, which puts its binaries into `bin/QUUX-default64` etc. 

When a GAP package containing an external executable (such as carat), it uses `DirectoriesPackagePrograms( "carat" )` to get a directory object pointing to the `bin/GARPACH-CONFIGNAME` directory inside the package; this is used to locate and launch the executables.

Now, the above code (or rather, its original version) does the following: Carat puts its binaries into `bin/GAPARCH`. It then proceds to create symlinks`bin/GAPARCH-CONFIGNAME` pointing to `bin/GAPARCH` for all configs it can find.

Your new doesn't do that anymore...

Comment (fingolfin) [inline bin/BuildPackages.sh]: Now, I would argue that the correct solution here is that GAP should allow packages to put executables (but not kernel extensions!) into a `bin/GAPARCH` directory anyway! Interestingly, there is code to do something like that in `DirectoriesPackagePrograms`! It will try both `GAPInfo.Architecture` (the full string `GAPARCH-CONFIGNAME`) but also `GAPInfo.ArchitectureBase`. Unfortunately, the latter is currently always equal to the former... I think this got addd and then later broken during GAP 4.5 development... I'll look into turning this into a "proper" general solution, and then we can get rid of this hack...

Comment (fingolfin) [inline bin/BuildPackages.sh]: I had a look at this. I forgot about one thing: GAP also inserts the compiler name into the directory name -- and carat matches that. But I don't think this makes sense for `DirectoriesPackagePrograms`.
So, this PR should not wait for any potential changes to `DirectoriesPackagePrograms`. Instead, just restore the original behavior of the script as I described it above.

Comment (fingolfin) [inline bin/BuildPackages.sh]: Hrm, but the old script also had a flaw: it was not idempotent.

And reading it, and looking at what carat does, I doubt the old script worked correctly: It does `ln -s "$aa" "`basename $x`"` where `aa` is the `config.guess` output -- so it does *not* contain the name of the C compiler. But the carat Makefile does insert that...

I'll try to write a proper replacement. Just got carat to compile, but it required quite some effort...

Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: The old script did _not_ work correctly, it made the wrong symlink, at least on my system. This currently compiles (I checked also the package's README), but yes, only with ````$GAParch````, and not with the other (potentially existing) directories. I liked your idea to use ````$GAParch````, so I did, and also because I liked it over parsing the output of ````ls```` or even ````find````. 

Anyway, just let me know how you want it to be changed, and I will do so.

**Comment by fingolfin** [comment]
I am still not 100% happy, but I think it's silly to block this PR (and in fact I am sorry that it took me so long to come to that realization). It is already a great improvement over what is in the repos. Merged now, more things can be added later.

Among the things I do not yet like is (as I already mentioned) that the logs contain color codes; and also that it simply discards unknown arguments. Thus it silently fails on various bad inputs like these:
```
$GAPROOT/bin/BuildPackages.sh --with-gaproot $GAPROOT # missing equal sign
$GAPROOT/bin/BuildPackages.sh io   # the package is named "io", but should have given the dir name
```
Anyway, that's all minor.
